Implement `asm.pseudo` for RISC-V

officialcjunior opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

What I'm hoping to add here, is support for `asm.pseudo` for the RISC-V architecture. I referred to [The RISC-V Instruction Set Manual](http://content.riscv.org/wp-content/uploads/2017/05/riscv-spec-v2.2.pdf), notably page 110, for the instructions.
As I am not an expert in this, I was mostly looking at the commit [e28ce](https://github.com/radareorg/radare2/commit/e28ce2990d90660a35053263555d796f034f74e3) and `parse_arm_pseudo.c` to make this possible.

**Test plan**
It seems to work, at least on layman's terms:

![image](https://user-images.githubusercontent.com/29057155/95674273-aab2a380-0bcc-11eb-8cc9-ca786b9b6883.png)

I've added a small test which succeeds, but I think its probably not enough, as it's not very big and I believe that there's lot more to fix and add in this PR.
```
$ r2r -i test/db/anal/riscv 
Loaded 4 tests.
Skipping json tests because jq is not available.
[4/4]                       4 OK         0 BR        0 XX        0 FX
Finished in 0 seconds.
```

I'm still learning how the parsing works, so here, I've basically just used the function `parse()` from `parse_arm_pseudo`, so it probably have many useless things, that I didn't understand or thought wasn't important. So, I'm hoping to fix them all, one by one, through your requests.

**Closing issues**
None

XVilka commented on libr/parse/p/parse_riscv_pseudo.c:
```diff
@@ -0,0 +1,229 @@
+				j = 0;
+				ch = ops[i].str[j];
+				for (j = 0, k = 0; ch != '\0'; j++, k++) {
+					ch = ops[i].str[j];
```
Too many levels of nesting - please split to a smaller blocks, functions, etc.

officialcjunior replied:
Thank you for your reply. I really appreciate it.

So, I was looking up about functions similar to what you are referring to and saw `can_replace()` in `parse_mips_pseudo` which I believe checks if something can be replaced. Is that what you're talking about here? Or is it the `if (ch == '#')` part?

If not, can you just write what and which exact part should be done by a function?

trufae replied:
he is refering to moving the body of the function to a separate function. so the whole function fits a single screen

trufae replied:
but as i said in my previous comment, i dont think refactoring this is really necessary because this code is duplicated multiple times and this interesting and useful feature may take ages to be merged if this cleanup ends in the full refactoring that implies having a common pseudo parser for all archs so we dont duplicate code

officialcjunior replied:
Marking as resolved.

XVilka commented on libr/parse/p/parse_riscv_pseudo.c:
```diff
@@ -0,0 +1,229 @@
+					memmove (num+1, ptr, strlen (ptr)+1);
+					ptr = (char*)r_str_lchr (buf, ']');
+					if (n && ptr) {
+						char *rest = strdup (ptr+1);
```
Same here.

officialcjunior replied:
And here, the stuff inside the `if (n && ptr)` should be made into a function?

trufae replied:
most of your code is copypasta from other pseudo plugins that are also made of pasta, we may probably want to implement generic parser, but all the new code should conform the rules.

trufae replied:
not needed,this block is small enough.

officialcjunior replied:
Okay, marking this as resolved

trufae commented on libr/parse/p/parse_riscv_pseudo.c:
```diff
@@ -0,0 +1,229 @@
+					if (n && ptr) {
+						char *rest = strdup (ptr+1);
+						if (n > 0) {
+							sprintf (ptr, "+%d]%s", n, rest);
```
dont use sprintf, use snprintf or so, otherwise this thing can overflow

trufae commented on libr/parse/p/parse_riscv_pseudo.c:
```diff
@@ -0,0 +1,229 @@
+				if (num) {
+					n = atoi (num+1);
+					*ptr = '[';
+					memmove (num+1, ptr, strlen (ptr)+1);
```
spaces around + and suhc..but you can use r_str_cpy, which is doing memmove under the hood, but you dont have to use the strlen by hand

officialcjunior commented:
Thanks for the reply.

Right now, I've used `r_str_cpy`, replaced `sprintf` with `snprintf` and added the missing spaces.

About `snprintf`, what should be the maximum number of bytes? 
I had used `sizeof (ptr)` here.

trufae commented:
Can you rebase the PR i dont see why it's not passing the CI otherwise

officialcjunior commented:
I think it's because I have not updated `libr/parse/meson.build`, yet.

Can I add that? I think that'll solve the issue.

Along with that, I saw `plugins.static.cfg`. Should I update that, too?

trufae commented:
Yep update those files

officialcjunior commented:
Did that, but we've got a problem with the `snprintf`: 
```
error: argument to ‘sizeof’ in ‘snprintf’ call is the same expression as the destination; did you mean to provide an explicit length? [-Werror=sizeof-pointer-memaccess]
        snprintf (ptr, sizeof (ptr), "+%d]%s", n, rest);
```
If you're around, you may suggest the changes, meanwhile I'll look up about fixing this.

trufae commented on libr/parse/p/parse_riscv_pseudo.c:
```diff
@@ -0,0 +1,229 @@
+					if (n && ptr) {
+						char *rest = strdup (ptr + 1);
+						if (n > 0) {
+							snprintf (ptr, sizeof (ptr), "+%d]%s", n, rest);
```
sizeof(ptr) is always size_t, so its not how you should use snprintf. 

* buf contents size len + 1
* ptr is buf+x

remaining chars in ptr are:

size_t remaining = len + 1 - (ptr - buf);

officialcjunior replied:
`snprintf (ptr, len + 1 - (ptr - buf), "%d]%s", n, rest);`
Like that?
I didn't produce any warnings.
What exactly is `(ptr - buf)`

trufae replied:
Better use a variable one line above to make it more readable.

This ptr-buf is the distance between the current char and the beginning of the buffer

officialcjunior replied:
Sure. Variable `dist`of type `int`?

trufae commented on libr/parse/p/parse_riscv_pseudo.c:
```diff
@@ -181,10 +181,11 @@ static int parse(RParse *p, const char *data, char *str) {
 					ptr = (char*)r_str_lchr (buf, ']');
 					if (n && ptr) {
 						char *rest = strdup (ptr + 1);
+						int dist = ptr - buf;
```
Why not hold the result in a variable instrad of computing it twice below?

Also. The type must be size_t

officialcjunior replied:
Alright, so, I'll make it:
`size_t dist = len + 1 - ptr - buf;`

trufae commented on libr/parse/p/riscv_pseudo.mk:
```diff
@@ -0,0 +1,16 @@
+	endif
+
+${TARGET_RISCVPSEUDO}: ${OBJ_RISCVPSEUDO}
+	${CC} $(call libname,parse_riscv_pseudo) -L../../util -llibr_util \
```
-llibr_util? It should be -lr_util

trufae reviewed (changes requested):
The failing test it seems to be because the parse is not working well and causes it to miss a var. can u compare the input and output of that string because this plugin is also in charge of doing that.. also its possible that the current test is wrong. So maybe also an option to update the output in the broken test. Because it was replaced using the wrrong parse plugin.

Btw did u checked if theres any other missing pseudo plugin for any of the archs in default r2?

trufae commented:
Please run `r2r -i test/db/formats/elf/elf-riscv64`

and press `f` to fix the test and commit the changes. this PR will also fix another arm64 travis test

eagleoflqj commented:
https://builds.sr.ht/~xvilka/job/332435#task-test
db/cmd/cmd_list also needs to be updated

trufae commented:
looks good to merge for me

officialcjunior commented:
Also @trufae, shall I replace `sprintf` with `snprintf` on `parse_att2intel` too?

trufae commented:
yep!
